cmd/forwarder/auth.go

```diff
@@ -0,0 +1,225 @@
+package main
+
+import (
+	"crypto/hmac"
+	"crypto/sha512"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-redis/redis"
+	metrics "github.com/rcrowley/go-metrics"
+	log "github.com/sirupsen/logrus"
+)
+
+func newAuth(config AuthConfig, registry metrics.Registry) (*BasicAuth, error) {
+	var err error
+
+	if config.RedisUrl != "" && config.RedisKey == "" {
+		err = errors.New("RedisKey must be set if RedisUrl is set")
+		return nil, err
+	}
+
+	if config.RedisUrl == "" && config.Tokens == "" {
+		err = errors.New("At least one of RedisUrl or Tokens must be set.")
+		return nil, err
+	}
+
+	result, err := NewBasicAuthFromString(config.Tokens, config.HmacKey, registry)
+
+	if err != nil {
+		return result, err
+	}
+
+	pChanges := metrics.GetOrRegisterCounter("log-iss.auth_refresh.changes", registry)
+	pFailures := metrics.GetOrRegisterCounter("log-iss.auth_refresh.failures", registry)
+	pSuccesses := metrics.GetOrRegisterCounter("log-iss.auth_refresh.successes", registry)
+
+	if config.RedisUrl != "" {
+		// Parse redis db out of url
+		u, err := url.Parse(config.RedisUrl)
+		if err != nil {
+			return result, err
+		}
+
+		var db int
+		if u.Path != "" && u.Path != "/" {
+			db, err = strconv.Atoi(u.Path[1:len(u.Path)])
+			if err != nil {
+				return result, err
+			}
+		}
+
+		ui := u.User
+		password := ""
+		if ui != nil {
+			password, _ = ui.Password()
+		}
+		client := redis.NewClient(
+			&redis.Options{
+				Addr:     u.Host,
+				Password: password,
+				DB:       db,
+			},
+		)
+
+		// Refresh once at the start.
+		// Ignore errors here - if redis is down, we don't want the process to fail to start.
+		_, _ = refreshAuth(result, client, config.HmacKey, config.RedisKey, config.Tokens)
+
+		// Refresh forever
+		ticker := time.NewTicker(config.RefreshInterval)
+		go func() {
```

Comment:
Suggestion: Consider making this whole thing a named function and the `go func(args)` that.